source/PlayState.hx

```diff
@@ -1,23 +1,64 @@
 package;
 
-import flixel.FlxSprite;
+import Wall;
+import flixel.FlxBasic;
+import flixel.FlxG;
+import flixel.FlxObject;
 import flixel.FlxState;
-import hero.Hero;
+import flixel.group.FlxGroup.FlxTypedGroup;
+import flixel.text.FlxText;
+import hero.ControllableHero;
+import hero.ScreensaverHero;
 
 class PlayState extends FlxState
 {
-	var hero:Hero;
+	var controllableHero:ControllableHero;
+	var walls:FlxTypedGroup<Wall>;
+	var screensaverHeroes:FlxTypedGroup<ScreensaverHero>;
+	var totalScore:Int = 0;
 
 	override public function create()
 	{
 		super.create();
 
-		hero = new Hero(300, 300);
-		add(hero);
+		screensaverHeroes = new FlxTypedGroup<ScreensaverHero>();
+		for (i in 0...30)
+		{
+			screensaverHeroes.add(new ScreensaverHero(300, 50));
+		}
+		add(screensaverHeroes);
+
+		var controllableHero = new ControllableHero();
+		add(controllableHero);
+
+		walls = new FlxTypedGroup<Wall>();
+		walls.add(new Wall(100, 100));
+		walls.add(new Wall(200, 100));
+		walls.add(new Wall(300, 100));
+		walls.add(new Wall(400, 100));
+		walls.add(new Wall(500, 100));
+		add(walls);
+
+		var text = new flixel.text.FlxText(75, 200, 0, " Hello World", 64);
+		text.setFormat(null, 64, 0x770088, CENTER, OUTLINE, 0xFF00FF00, true);
+		text.borderSize = 5;
+		add(text);
+	}
+
+	private function AddScore(obj1:flixel.FlxBasic, obj2:flixel.FlxBasic)
+	{
+		obj1.kill();
+		totalScore += 1;
+		var score = new flixel.text.FlxText(0, 0, 0, "Score: " + totalScore, 12, true);
+		add(score);
 	}
 
 	override public function update(elapsed:Float)
 	{
+		// FlxG.overlap(screensaverHeroes, controllableHero, AddScore(screensaverHeroes, controllableHero));
+		// couldnt figure out how to make this part work. ^
```

Comment:
There were two problems going on here, long-winded explanation for one of them follows.

Overlap's function signature looks like this:
```
overlap(?ObjectOrGroup1:FlxBasic, ?ObjectOrGroup2:FlxBasic,  ?NotifyCallback:(Dynamic, Dynamic) ‑> Void,  ?ProcessCallback:(Dynamic, Dynamic) ‑> Void):Bool
```

it's not very easy to digest, unfortunately, but we can break down the arguments to the function piece-by-piece to help make it more understandable. Our 4 parameters for the function are ObjectOrGroup1, ObjectOrGroup2, NotifyCallback, ProcessCallback.

`ObjectOrGroup1` and `ObjectOrGroup2` are pretty straightforward, they're the things that we want to compare against one another to check for overlaps.

The `NotifyCallback` and `ProcessCallback` are pretty funky looking, but they're actually not so bad. We know that in Haxe, we define parameters for a function like this:
```
<ParameterName>:<ParameterType>
```
So if we define a function 
```
function displayTwoNumbers(num1:Int, num2:Float)
```
We're saying that `displayTwoNumbers` is a function that takes two parameters, `num1` which is an `Int`, and `num2` which is a `Float`.

So, let's figure out NotifyCallback:
```
NotifyCallback:(Dynamic, Dynamic) -> Void
```
Using that colon in the middle as our guide, we can say that this is a parameter named `NotifyCallback`, which is of the type `(Dynamic, Dynamic) -> Void`, whatever that means.

By digging into [Haxe's documentation](https://haxe.org/manual/types-function.html), we can find out that this weird notation is actually how Haxe displays a function's type. `(Dynamic, Dynamic) -> Void` means: a function that has two `Dynamic`-typed objects as its parameters, then returns `Void` (nothing).

So what our Overlap function wants `NotifyCallback` to be is an uncalled function that has two parameters of any type (`Dynamic` is like a wild-card "any type goes" sort of thing).

What you've got commented out
```
FlxG.overlap(screensaverHeroes, controllableHero, AddScore(screensaverHeroes, controllableHero));
```
almost looks like it should work - AddScore is the function you wanted to trigger whenever the `screensaverHeros` group and `controllableHero` object run into each other. The problem is that by putting `AddScore(screensaverHeroes, controllableHero)` you're actually calling the `AddScore` function exactly once when this line executes, then passing the *result* of that call into `FlxG.overlap`.

Any time that you put parenthesis after a function name, you're telling the program to execute that method right away and to use the result of that call in any further computations. That's why something like `walls.add(new Wall(100, 100));` earlier in this file works out okay - `new Wall(100, 100)` is evaluated first, the result of that function is a new `Wall` object which is then passed into `walls.add()`

So in this case putting 
```
FlxG.overlap(screensaverHeroes, controllableHero, AddScore(screensaverHeroes, controllableHero));
```
would end up evaluating to
```
FlxG.overlap(screensaverHeroes, controllableHero, Void);
```
Since your `AddScore()` method doesn't return anything when it's called.

Instead, you want to do
```
FlxG.overlap(screensaverHeroes, controllableHero, AddScore);
```
which lets the overlap method get the `screensaverHeroes`, the `controllableHero`, and the `AddScore` function all passed in as it wants - now it can reuse that AddScore function over and over again as compares the objects against each other.

Reply:
The other problem is that controllable hero actually isn't set when this code runs, so you end up with
```
FlxG.overlap(screensaverHeroes, null, AddScore);
```
And a weird quirk of `overlap` is that if you pass in a group for the first parameter and null in for the second, it just assumes that you're trying to collide all of the objects in `screensaverHeroes` against one another. See comment on line 31 to fix.